File: python/src/wdl_model/model/wdl_version.py
"""Supported WDL language versions for the Python object model and validators."""

from __future__ import annotations

from enum import Enum


class WdlVersion(Enum):
    """Enumeration of the WDL versions supported by this library."""

    V1_0 = (1, 0, "1.0")
    V1_1 = (1, 1, "1.1")
    V1_2 = (1, 2, "1.2")
    V1_3 = (1, 3, "1.3")

    def __init__(self, major: int, minor: int, versionString: str):
        self.major = major
        self.minor = minor
        self.versionString = versionString

    def getMajor(self) -> int:
        """Return the major version component."""
        return self.major

    def getMinor(self) -> int:
        """Return the minor version component."""
        return self.minor

    def getVersionString(self) -> str:
        """Return the source-level WDL version string, for example ``1.3``."""
        return self.versionString

    @classmethod
    def fromString(cls, versionString: str) -> "WdlVersion":
        """Convert a source-level version string into the matching enum value."""
        for item in cls:
            if item.versionString == versionString:
                return item
        raise ValueError(f"Unsupported WDL version: {versionString}")

File: typescript/src/model/wdl-version.ts
/** Supported WDL language versions. */
import type { WdlNode } from './base/wdl-node.js';

/** Models the supported WDL version set used by the parser and validators. */
export class WdlVersion implements WdlNode {
  public static readonly V1_0 = new WdlVersion(1, 0, '1.0');
  public static readonly V1_1 = new WdlVersion(1, 1, '1.1');
  public static readonly V1_2 = new WdlVersion(1, 2, '1.2');
  public static readonly V1_3 = new WdlVersion(1, 3, '1.3');

  /** Returns all supported version values. */
  public static values(): readonly WdlVersion[] {
    return [WdlVersion.V1_0, WdlVersion.V1_1, WdlVersion.V1_2, WdlVersion.V1_3];
  }

  /** Converts a source-level version string such as `1.3` into the matching model value. */
  public static fromString(versionString: string): WdlVersion {
    const match = WdlVersion.values().find((value) => value.versionString === versionString);
    if (!match) {
      throw new Error(`Unsupported WDL version: ${versionString}`);
    }
    return match;
  }

  /** Creates a version value from its major/minor components and display string. */
  private constructor(
    public readonly major: number,
    public readonly minor: number,
    public readonly versionString: string,
  ) {}

  /** Returns the source-level version string. */
  public getVersionString(): string {
    return this.versionString;
  }
}
